docs/basics/metadata.md

```diff
@@ -0,0 +1,222 @@
+
+
+The **required** keys include:
+    - `name`: The name of the smart contract.
+    - `version`: The version of the smart contract.
+    - `authors`: The authors of the smart contract.
+
+It can _optionally_ include the following keys:
+    - `description`: The description of the smart contract.
+    - `documentation`: Link to the documentation of the smart contract.
+    - `repository`: Link to the code repository of the smart contract.
+    - `homepage`: Link to the homepage of the smart contract.
+    - `license`: The license of the smart contract.
+
+```json
+"contract": {
+  "name": "flipper",
+  "version": "3.1.0",
+  "authors": [
+    "Parity Technologies <admin@parity.io>"
+  ]
+}
+```
+
+TODO: Code reference from `cargo-contract`: https://github.com/paritytech/cargo-contract/blob/45fbc0b43ac9e676107ad9a03f8d7a0a01d84c50/metadata/lib.rs#L395
+
+### ABI
+This is the specification of the contract. The "outer" metadata format does not dicate how
+this looks, and instead will be driven by each smart contracting language. In this
+document we will focus on the ink! ABI.
+
+The ink! metadata consists of the following **required** sections
+    - `spec`: The description of the contract (e.g constructors, messages, events, etc.).
+    - `storage`: The layout of the storage data structure
+    - `types`: A read-only registry containing types in their portable form for
+      serialization.
+
+#### `spec`
+The contract `spec` consists of the following **required** keys:
+    - `constructors`: The set of constructors of the contract.
+        - `label`: The label of the constructor. In case of a trait provided constructor
+          the label is prefixed with the trait label.
+        - `selector`: The selector hash of the message.
+        - `payable`: If the constructor accepts any `value` from the caller.
+        - `args`: The parameters of the deployment handler.
+        - `docs`: The deployment handler documentation.
+    - `messages`: The external messages of the contract.
+        - `label`: The label of the message. In case of trait provided messages and
+          constructors the prefix by convention in ink! is the label of the trait.
+        - `selector`: The selector hash of the message.
+        - `mutates`: If the message is allowed to mutate the contract state.
+        - `payable`: If the message accepts any `value` from the caller.
+        - `args`: The parameters of the message.
+        - `return_type`: The return type of the message.
+        - `docs`: The message documentation.
+    - `events`: The events of the contract.
+        - `label`: The label of the event.
+        - `args`: The event arguments.
+        - `docs`: The event documentation.
+    - `docs`: The contract documentation.
+
+Note that while all these keys are required, they may be empty. For example, if a
+contract does not define any events then the `events` key would contain an empty array
+`[]`.
+
+#### `storage`
```

Comment:
Is this documenting the existing or the new storage layout metadata which will be populated my Green's storage rework PR? It should document the new format (or leave it out until that PR is merged), not sure how different it is